File: NCPC/2017/ncpc2017-packages/importspaghetti/submissions/accepted/per.cc
#include <bits/stdc++.h>
using namespace std;

map<string, int> name_idx;
string names[1501];
int dst[1501][1501];
int n;

int lookup(string name) {
    int &res = name_idx[name];
    if (!res) {
        res = name_idx.size();
        names[res-1] = name;
    }
    return res-1;
}

void printpath(int u, int v) {
    for (int i = 0; i < n; ++i)
        if (dst[u][i] == 1 && (dst[i][v] == dst[u][v] - 1 || dst[u][v] == 1)) {
            printf("%s ", names[u].c_str());
            if (dst[u][v] > 1)
                printpath(i, v);
            break;
        }
}

int main(void) {
    string str;
    cin >> n;
    getline(cin, str);
    getline(cin, str);
    memset(dst, 0x1f, sizeof(dst));
    for (int i = 0; i < n; ++i) {
        int k;
        cin >> str >> k;
        int u = lookup(str);
        getline(cin, str);
        for (int i = 0; i < k; ++i) {
            getline(cin, str);
            istringstream in(str);
            in >> str;
            while (in >> str) {
                if (str.back() == ',') str.pop_back();
                dst[u][lookup(str)] = 1;
            }
        }
    }
    for (int k = 0; k < n; ++k)
        for (int i = 0; i < n; ++i)
            for (int j = 0; j < n; ++j)
                dst[i][j] = min(dst[i][j], dst[i][k] + dst[k][j]);
    int s = 0;
    for (int i = 0; i < n; ++i)
        if (dst[i][i] < dst[s][s])
            s = i;
    if (dst[s][s] <= n)
        printpath(s, s);
    else
        printf("SHIP IT\n");
}

File: NCPC/2017/ncpc2017-packages/importspaghetti/submissions/accepted/per.py
def printpath(u, v):
    for i in range(n):
        if dst[u][i] == 1 and (dst[i][v] == dst[u][v] - 1 or dst[u][v] == 1):
            print names[u]
            if dst[u][v] > 1:
                printpath(i, v);
            break

n = int(raw_input())
names = raw_input().split()
idx = dict(zip(names, range(n)))
dst = [[n+1]*n for _ in range(n)]

for i in range(n):
    (_, k) = raw_input().split()
    for _ in range(int(k)):
        for imp in raw_input().split()[1:]:
            if imp[-1] == ',': imp = imp[:-1]
            dst[i][idx[imp]] = 1

V = range(n)
for k in V:
    dk = dst[k]
    for i in V:
        di = dst[i]
        for j in V:
            di[j] = min(di[j], di[k] + dk[j]);

s = 0
for i in range(n):
    if dst[i][i] < dst[s][s]:
        s = i
if dst[s][s] <= n:
    printpath(s, s)
else:
    print "SHIP IT"
